restore_xlog.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH RESTORE_XLOG 1 "October 11, 2015"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
restore_xlog \- retrieve WAL segment files during database recovery

.SH SYNOPSIS
.B restore_xlog
.RI [ options ]
.I xlogfile destination


.SH DESCRIPTION
The \fBrestore_xlog\fP script is the default action that is used for the
\fBpostgres\fP(1) \fIrestore_command\fP in \fIrecovery.conf\fP by
\fBpitrery\ restore\fP.  It is responsible for retrieving the archived WAL
segments that are needed during the recovery operation.


.SH OPTIONS
The following options are available:

.TP
.BI "\-C " conf
Use option settings from the specified configuration file instead of the
default \fIpitrery.conf\fP file.  This may be a full path to the file to use,
or the name of a configuration file in \fI/etc/pitrery\fP.

.TP
.BI "\-a " [[user@]host:]/dir
The directory on the (local or remote) host where WAL segment files will be
stored.  This will override the \fBARCHIVE_DIR\fP option from the
configuration file.  When the host name is omited, the archiving is local.

.TP
.B \-X
Retrieve WAL segment files that are not compressed.  This will override the
\fBARCHIVE_COMPRESS\fP option from the configuration file (forcing it to be
"no").

.TP
.BI "\-c " command
The command to use for uncompressing WAL segment files.  This will override
the \fBARCHIVE_UNCOMPRESS_BIN\fP option from the configuration file.

.TP
.BI "\-s " suffix
The compressed file suffix to expect (e.g., gz, bz2, xz).  This will override
the \fBARCHIVE_COMPRESS_SUFFIX\fP option from the configuration file.

.TP
.B \-S
Send messages to the \fBsyslog\fP(3) instead of \fBstdio\fP(3).  This will
override the \fBSYSLOG\fP option from the configuration file (forcing it to
be "yes").  Messages sent to \fIstdout\fP will be logged with \fBLOG_INFO\fP
priority, while messages to \fIstderr\fP will be logged with \fBLOG_ERR\fP
priority.

.TP
.BI "\-f " facility
Specify the \fBsyslog\fP facility to use.  This will override the
\fBSYSLOG_FACILITY\fP option from the configuration file.  Defaults to use
\fBlocal0\fP if not otherwise explicitly set.  See \fBlogger\fP(1) for details
of the valid facility strings that can be used here.

.TP
.BI "\-t " ident
An identifier to prefix \fBsyslog\fP output with.  This will override the
\fBSYSLOG_IDENT\fP option from the configuration file.  Defaults to use the
string "postgres" if not otherwise explicitly set.

.TP
.B \-T
Timestamp the log messages.  This will override the \fBLOG_TIMESTAMP\fP option
from the configuration file (forcing it to be "yes").

.TP
.B \-?
Output the brief help summarising the above options.

.TP
.I xlogfile
The name of the WAL segment to restore, relative to the configured
archive directory. It must be set to \fB%f\fP inside
\fIrecovery.conf\fP when configuring restore_command.

.TP
.I destination
The target path where to restore the file. It must be set to \fB%p\fP
inside \fIrecovery.conf\fP when configuring restore_command.

.SH FILES
.TP
.I /etc/pitrery/pitrery.conf
The default \fBrestore_xlog\fP configuration file if not explicitly specified.

.TP
.I $PGATA/recovery.conf
The file created in the \fBPGDATA\fP directory by \fBpitrery\ restore\fP which
configures if and/or how to use \fBrestore_xlog\fP to retrieve any needed WAL
segment files during database recovery.


.SH SEE ALSO
.BR archive_xlog (1),
.BR pitrery (1).


.SH AUTHOR
.B restore_xlog
was written by Nicolas Thauvin <nicolas.thauvin@dalibo.com>.
This man page was written by Ron <ron@debian.org>.



CHANGELOG
pitrery 2.0
===========

  * Regroup all action into the main script

  * Change the sed commands in the Makefile to have the script work
    without modifications

  * Command line switches and options to specify if the backup is
    local, the user, host and target directory, are now merged into a
    SSH style syntax: `[[user@]host:]/path`. Not providing a host
    tells pitrery the backups are local.

  * Rename the default configuration file from `pitr.conf` to `pitrery.conf`

  * Comment out all parameters in the configuration file so that they
    do not interfere

  * Remove the BACKUP_LABEL option, subdirectory and -l switch. They
    were not used.

  * Support for PostgreSQL 10


pitrery 1.13
============

  * Fix tablespaces links being overwritten by postmaster on 9.5 due
    to tablespace_map. This conflicted with our tablespace relocation

  * Fix the purge of WAL file that would remove all archives if the
    stop file is missing

  * Fix backups failing due to incorrect parsing of the stop time
    when iso8601 timestamps are used


pitrery 1.12
============

  * New -m option for archive_xlog to chmod the archived file to the
    specified octal permission bits

  * Fix the restore so that the backup_label is restore whan used with
    PostgreSQL 9.6

  * Remove support of deprecated compression parameters, renamed in 1.9


pitrery 1.11
============

  * New configure action to create a configuration file from the
    command line.

  * New check action to test configuration files.

  * Do not backup replication slots state files and produce an SQL
    file to recreate them after restore.

  * Add a dry run option (-N) for purge.

  * Improve the quoting and ensure unusual characters are handled
    properly.

  * Add manpages for the user facing commands, pitrery, archive_xlog
    and restore_xlog.

  * Relax restrictions on the target date format on restore. It was
    meant to have a working format common to all free unices
    platforms, but the user can just adapt the command to its
    system. So now one can do run: pitrery restore -d '1 hour ago', or
    anything the local date command can parse.

  * New ARCHIVE_OVERWRITE configuration parameter along with -O option
    (do not overwrite) for archive_xlog to check if the target files
    exists before archiving, when set to "no" or -O is on the command
    line.

  * New ARCHIVE_CHECK (-H) and ARCHIVE_FLUSH (-F) for archive_xlog. The
    first one check md5 sum of the archived WAL file, the second force
    a flush to disk.

  * Make use of the new backup API of PostgreSQL 9.6

  * Optionally name backups dir with ISO 8601 timestamps, with config
    option USE_ISO8601_TIMESTAMPS and command line switch -t for the
    backup action.

  * French translation of the documentation.

  * Bugfixes and better handling of unusual characters in paths


pitrery 1.10
============

  * Fix issue in restore_xlog not using SSH user and host found in the
    configuration file, thus preventing restore to work out of the
    box.

  * Ensure archive_xlog create the target directory on remote hosts

  * Only purge WAL files when there are some WAL files to remove


pitrery 1.9
===========

  * Do not exit without error when the server is in recovery. This is
    not a consistent behaviour.

  * Improve the speed of the purge of archives WAL files by stopping
    the purge when the limit is reached.

  * Add the -T switch along with the LOG_TIMESTAMP option to prefix
    messages of backup, restore and purge with the date and time.

  * Rename the compression configuration parameters for WAL files to
    avoid confusion :
    * `COMPRESS_BIN` -> `ARCHIVE_COMPRESS_BIN`
    * `COMPRESS_SUFFIX` -> `ARCHIVE_COMPRESS_SUFFIX`
    * `UNCOMPRESS_BIN` -> `ARCHIVE_UNCOMPRESS_BIN`

  * Add the -c and -s switch to archive_xlog and restore_xlog which
    allow to give the (un)compression command and compressed file suffix
    respectively.

  * No longer use command line switch to configure restore_xlog for
    restore_command in restore action. Give a configuration file or
    the full command to use. Thus only -r and -C (changed to give the
    configuration file) remains for archived WAL options.

  * Add the -l switch to pitrery. It lists the configuration files
    inside the default configuration directory.

  * Include all recovery parameters commented out in the recovery.conf
    generated at restore.


pitrery 1.8
===========

  * Exclude temporary locations from backup

  * Support pax to create hardlinks, for better portability. GNU cp
    stays the default, the configuration goes to the Makfile.

  * Add a -V switch to the pitrery wrapper script to show the version.

  * Add an option (-R) to overwrite target directory on restore.


pitrery 1.7
===========

  * Show the storage method and compression suffix when listing
    backups in verbose mode.

  * Allow configuration of compression tools used when storing backup
    contents using tar.

  * Remove the `pitr_mgr` symlink.

  * Have `archive_xlog` create the local target directory so that we don't
    end up archiving on the same path, when the user forget to create it.

  * Run backup hook command even if the backup fails. The exit code of the
    backup is available to the hook in the PITRERY_EXIT_CODE env variable.

  * Bugfix on sed -E vs sed -r, use the later for older GNU sed.


pitrery 1.6
===========

  * Exit without error when backup is run on a hot standby
    server. Issue a warning message when run interactively, so an
    active cron job does not send any mail.

  * Bugfixes and improvements on documentation


pitrery 1.5
===========

  * Force the checkpoint on pg_start_backup() when possible

  * Support rsync for the base backup. tar remains the default with a
    new -s option for backup_pitr and the STORAGE configuration
    variable

  * Save configuration files if they are outside of PGDATA

  * Rename `pitr_mgr` to `pitrery`, with a symlink for backward

restore_xlog.bash
#!@BASH@
#
# Copyright 2011-2017 Nicolas Thauvin. All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:
# 
#  1. Redistributions of source code must retain the above copyright
#     notice, this list of conditions and the following disclaimer.
#  2. Redistributions in binary form must reproduce the above copyright
#     notice, this list of conditions and the following disclaimer in the
#     documentation and/or other materials provided with the distribution.
# 
# THIS SOFTWARE IS PROVIDED BY THE AUTHORS ``AS IS'' AND ANY EXPRESS OR
# IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
# OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
# IN NO EVENT SHALL THE AUTHORS OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
# INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
# (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
# ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
# THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
#

# Apply an extra level of shell quoting to each of the arguments passed.
# This is necessary for remote-side arguments of ssh (including commands that
# are executed by the remote shell and remote paths for scp and rsync via ssh)
# since they will strip an extra level of quoting off on the remote side.
# This makes it safe for them to include spaces or other special characters
# which should not be interpreted or cause word-splitting on the remote side.
qw() {
    while (( $# > 1 )); do
	printf "%q " "$1"
	shift
    done
    (( $# == 1 )) && printf "%q" "$1"
}

# Message functions
now() {
    [ "$LOG_TIMESTAMP" = "yes" ] && echo "$(date "+%F %T %Z ")"
}

error() {
    echo "$(now)ERROR: $*" 1>&2
    exit 1
}

warn() {
    echo "$(now)WARNING: $*" 1>&2
}

# Script help
usage() {
    echo "usage: `basename $0` [options] xlogfile destination"
    echo "options:"
    echo "    -L             restore local archives"
    echo "    -C conf        configuration file"
    echo "    -u username    username for SSH login"
    echo "    -h hostname    hostname for SSH login"
    echo "    -d dir         directory containing WALs on host"
    echo "    -X             do not uncompress"
    echo "    -c command     uncompression command"
    echo "    -s suffix      compressed file suffix (ex: gz)"
    echo "    -S             send messages to syslog"
    echo "    -f facility    syslog facility"
    echo "    -t ident       syslog ident"
    echo
    echo "    -T             Timestamp log messages"
    echo "    -?             print help"
    echo
    exit $1
}

# Default configuration
CONFIG=pitr.conf
CONFIG_DIR=@SYSCONFDIR@
ARCHIVE_DIR=/var/lib/pgsql/archived_xlog
ARCHIVE_LOCAL="no"
SYSLOG="no"
ARCHIVE_COMPRESS="yes"
ARCHIVE_UNCOMPRESS_BIN=gunzip
ARCHIVE_COMPRESS_SUFFIX=gz

# CLI processing
while getopts "LC:u:h:d:Xc:s:Sf:t:?" opt; do
    case $opt in
	L) CLI_ARCHIVE_LOCAL="yes";;
	C) CONFIG=$OPTARG;;
	u) CLI_ARCHIVE_USER=$OPTARG;;
	h) CLI_ARCHIVE_HOST=$OPTARG;;
	d) CLI_ARCHIVE_DIR=$OPTARG;;
	X) CLI_ARCHIVE_COMPRESS="no";;
	c) CLI_ARCHIVE_UNCOMPRESS_BIN=$OPTARG;;
	s) CLI_ARCHIVE_COMPRESS_SUFFIX=$OPTARG;;
	S) CLI_SYSLOG="yes";;
	f) CLI_SYSLOG_FACILITY=$OPTARG;;
	t) CLI_SYSLOG_IDENT=$OPTARG;;
	T) CLI_LOG_TIMESTAMP="yes";;
	"?") usage 1;;
	*) error "Unknown error while processing options";;
    esac
done

# Check if the config option is a path or just a name in the
# configuration directory.  Prepend the configuration directory and
# .conf when needed.
if [[ $CONFIG != */* ]]; then
    CONFIG="$CONFIG_DIR/$(basename -- "$CONFIG" .conf).conf"
fi

# Load configuration file
if [ -f "$CONFIG" ]; then
    . "$CONFIG"

    # Check for renamed parameters between versions
    if [ -n "$UNCOMPRESS_BIN" ] && [ -z "$CLI_ARCHIVE_UNCOMPRESS_BIN" ]; then
	ARCHIVE_UNCOMPRESS_BIN=$UNCOMPRESS_BIN
	warn "restore_xlog: UNCOMPRESS_BIN is deprecated. please use ARCHIVE_UNCOMPRESS_BIN."
    fi
    if [ -n "$COMPRESS_SUFFIX" ] && [ -z "$CLI_ARCHIVE_COMPRESS_SUFFIX" ]; then
	ARCHIVE_COMPRESS_SUFFIX=$COMPRESS_SUFFIX
	warn "restore_xlog: COMPRESS_SUFFIX is deprecated. please use ARCHIVE_COMPRESS_SUFFIX."
    fi
fi

# Override configuration with cli options
if [ -n "$CLI_ARCHIVE_HOST" ]; then
    ARCHIVE_HOST=$CLI_ARCHIVE_HOST
    [ -n "$CLI_ARCHIVE_USER" ] && ARCHIVE_USER=$CLI_ARCHIVE_USER
    # When a host storing the archives is given for local to no, as it
    # can come from the configuration file
    ARCHIVE_LOCAL="no"
fi
[ -n "$CLI_ARCHIVE_LOCAL" ] && ARCHIVE_LOCAL=$CLI_ARCHIVE_LOCAL
[ -n "$CLI_ARCHIVE_DIR" ] && ARCHIVE_DIR=$CLI_ARCHIVE_DIR
[ -n "$CLI_ARCHIVE_COMPRESS" ] && ARCHIVE_COMPRESS=$CLI_ARCHIVE_COMPRESS
[ -n "$CLI_ARCHIVE_UNCOMPRESS_BIN" ] && ARCHIVE_UNCOMPRESS_BIN=$CLI_ARCHIVE_UNCOMPRESS_BIN
[ -n "$CLI_ARCHIVE_COMPRESS_SUFFIX" ] && ARCHIVE_COMPRESS_SUFFIX=$CLI_ARCHIVE_COMPRESS_SUFFIX
[ -n "$CLI_SYSLOG" ] && SYSLOG=$CLI_SYSLOG
[ -n "$CLI_SYSLOG_FACILITY" ] && SYSLOG_FACILITY=$CLI_SYSLOG_FACILITY
[ -n "$CLI_SYSLOG_IDENT" ] && SYSLOG_IDENT=$CLI_SYSLOG_IDENT
[ -n "$CLI_LOG_TIMESTAMP" ] && LOG_TIMESTAMP=$CLI_LOG_TIMESTAMP

# Redirect output to syslog if configured
if [ "$SYSLOG" = "yes" ]; then
    SYSLOG_FACILITY=${SYSLOG_FACILITY:-local0}
    SYSLOG_IDENT=${SYSLOG_IDENT:-postgres}

    exec 1> >(logger -t "$SYSLOG_IDENT" -p "${SYSLOG_FACILITY}.info")
    exec 2> >(logger -t "$SYSLOG_IDENT" -p "${SYSLOG_FACILITY}.err")
fi

# Check input: the name of the xlog file (%f) is needed as well has the target path (%p)
# PostgreSQL gives those two when executing restore_command
xlog=${@:$OPTIND:1}
target_path=${@:$(($OPTIND+1)):1}

if [ -z "$xlog" ] || [ -z "$target_path" ]; then
    error "missing xlog filename and/or target path. Please use %f and %p in restore_command"
fi

# Check if we have enough information on where to get the file
if [ "$ARCHIVE_LOCAL" != "yes" ] && [ -z "$ARCHIVE_HOST" ]; then
    error "Could not find where to get the file from (local or ssh?)"
fi

if [ "$ARCHIVE_LOCAL" = "yes" ] && [ -n "$ARCHIVE_HOST" ]; then
    error "ARCHIVE_LOCAL and ARCHIVE_HOST are set, it can't be both"
fi

# the filename to retrieve depends on compression
if [ "$ARCHIVE_COMPRESS" = "yes" ]; then
    xlog_file=${xlog}.$ARCHIVE_COMPRESS_SUFFIX
    target_file=${target_path}.$ARCHIVE_COMPRESS_SUFFIX
else
    xlog_file=$xlog
    target_file=$target_path
fi

# Get the file: use cp when the file is on localhost, scp otherwise
if [ "$ARCHIVE_LOCAL" = "yes" ]; then
    if [ -f "$ARCHIVE_DIR/$xlog_file" ]; then
	if ! cp -- "$ARCHIVE_DIR/$xlog_file" "$target_file"; then
	    error "could not copy $ARCHIVE_DIR/$xlog_file to $target_file"
	fi
    else
	error "could not find $ARCHIVE_DIR/$xlog_file"
    fi
else
    # check if we have a IPv6, and put brackets for scp
    echo $ARCHIVE_HOST | grep -qi '^[0123456789abcdef:]*:[0123456789abcdef:]*$' && ARCHIVE_HOST="[${ARCHIVE_HOST}]"

    if ! scp -- "${ARCHIVE_USER:+$ARCHIVE_USER@}$ARCHIVE_HOST:$(qw "$ARCHIVE_DIR/$xlog_file")" "$target_file" >/dev/null; then
	error "could not copy $ARCHIVE_HOST:$ARCHIVE_DIR/$xlog_file to $target_file"
    fi
fi

# Uncompress the file if needed
if [ "$ARCHIVE_COMPRESS" = "yes" ]; then
    if ! $ARCHIVE_UNCOMPRESS_BIN "$target_file"; then
	error "could not uncompress $target_file"
    fi
fi

exit 0
